snapcraft/extensions/env_injector.py

```diff
@@ -0,0 +1,129 @@
+
+    This extension allows you to transform snap options into environment
+    variables
+
+    It configures your application to run a command-chain that transforms the
+    snap options into environment variables automatically.
+
+    - To set global environment variables for all applications **inside** the snap:
+
+    .. code-block:: shell
+        sudo snap set <snap-name> env.<key>=<value>
+
+    - To set environment variables for a specific application **inside** the snap:
+
+    .. code-block:: shell
+        sudo snap set <snap-name> apps.<app-name>.env.<key>=<value>
+
+    - To set environment file inside the snap:
+
+    .. code-block:: shell
+        sudo snap set <snap-name> env-file=<path-to-env-file>
+    - To set environment file for a specific app:
+
+    .. code-block:: shell
+        sudo snap set <snap-name> apps.<app-name>.envfile=<path-to-env-file>
+
+    """
+
+    @staticmethod
+    @overrides
+    def get_supported_bases() -> Tuple[str, ...]:
+        return ("core24",)
+
+    @staticmethod
+    @overrides
+    def get_supported_confinement() -> Tuple[str, ...]:
+        return ("strict", "devmode", "classic")
+
+    @staticmethod
+    @overrides
+    def is_experimental(base: Optional[str]) -> bool:
+        return True
+
+    @overrides
+    def get_root_snippet(self) -> Dict[str, Any]:
+        return {}
+
+    @overrides
+    def get_app_snippet(self, *, app_name: str) -> Dict[str, Any]:
+        """Return the app snippet to apply."""
+        return {
+            "command-chain": ["bin/command-chain/env-exporter"],
+            "environment": {
+                "env_alias": f"{app_name}",
+            },
+        }
+
+    @overrides
+    def get_part_snippet(self, *, plugin_name: str) -> Dict[str, Any]:
+        return {}
+
+    @overrides
+    def get_parts_snippet(self) -> Dict[str, Any]:
+        toolchain = self.get_toolchain(self.arch)
+        if toolchain is None:
+            raise ValueError(
+                f"Unsupported architecture for env-injector extension: {self.arch}"
+            )
+
+        return {
+            "env-injector/env-injector": {
+                "source": f"{get_extensions_data_dir()}/env-injector",
+                "plugin": "nil",
+                "build-snaps": ["rustup"],
+                "build-packages": ["upx-ucl"],  # for binary compression
+                "override-build": f"""
+                rustup default stable
+                rustup target add {toolchain}
+
+                cargo build --target {toolchain} --release
+                mkdir -p $SNAPCRAFT_PART_INSTALL/bin/command-chain
+
+                cp target/{toolchain}/release/env-exporter $SNAPCRAFT_PART_INSTALL/bin/command-chain
+
+                # compress the binary
+                upx --best --lzma target/{toolchain}/release/env-exporter
+                cp target/{toolchain}/release/env-exporter $SNAPCRAFT_PART_INSTALL/bin/command-chain/env-exporter-upx
```

Comment:
This is actually copying both the original and compressed binary to the snap, where only the original one is used in the command chain on line 79.

```suggestion
                # compress the binary
                upx --best --lzma target/{toolchain}/release/env-exporter
                
                cp target/{toolchain}/release/env-exporter $SNAPCRAFT_PART_INSTALL/bin/command-chain
```